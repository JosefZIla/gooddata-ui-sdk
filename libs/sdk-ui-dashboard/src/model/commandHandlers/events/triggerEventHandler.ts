// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { TriggerEvent } from "../../commands";
import { DashboardEvents, ICustomDashboardEvent } from "../../events";
import { dispatchDashboardEvent } from "../../state/_infra/eventDispatcher";

export function* triggerEventHandler(ctx: DashboardContext, cmd: TriggerEvent): SagaIterator<void> {
    // fill the ctx property of the event properly so that the caller does not need to do this
    const fullEvent: DashboardEvents | ICustomDashboardEvent = {
        ...cmd.payload.eventBody,
        ctx,
    };

    yield dispatchDashboardEvent(fullEvent);
}
