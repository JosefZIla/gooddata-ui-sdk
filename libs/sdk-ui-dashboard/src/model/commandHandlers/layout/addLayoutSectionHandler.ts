// (C) 2021 GoodData Corporation

import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { AddLayoutSection } from "../../commands";
import { invalidArgumentsProvided } from "../../events/general";
import { selectLayout, selectStash } from "../../state/layout/layoutSelectors";
import { call, put, SagaReturnType, select } from "redux-saga/effects";
import { ExtendedDashboardLayoutSection } from "../../types/layoutTypes";
import isEmpty from "lodash/isEmpty";
import { layoutActions } from "../../state/layout";
import { DashboardLayoutSectionAdded, layoutSectionAdded } from "../../events/layout";
import { validateSectionPlacement } from "./validation/layoutValidation";
import { ItemResolutionResult, validateAndResolveStashedItems } from "./validation/stashValidation";
import { resolveIndexOfNewItem } from "../../utils/arrayOps";
import { selectInsightsMap } from "../../state/insights/insightsSelectors";
import { batchActions } from "redux-batched-actions";
import { insightsActions } from "../../state/insights";
import {
    validateAndNormalizeWidgetItems,
    validateAndResolveItemFilterSettings,
} from "./validation/itemValidation";

type AddLayoutSectionContext = {
    readonly ctx: DashboardContext;
    readonly cmd: AddLayoutSection;
    readonly layout: ReturnType<typeof selectLayout>;
    readonly stash: ReturnType<typeof selectStash>;
    readonly availableInsights: ReturnType<typeof selectInsightsMap>;
};

function validateAndResolveItems(commandCtx: AddLayoutSectionContext): ItemResolutionResult {
    const {
        ctx,
        layout,
        stash,
        cmd: {
            payload: { index, initialItems = [] },
        },
    } = commandCtx;

    if (!validateSectionPlacement(layout, index)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to insert new section at wrong index ${index}. There are currently ${layout.sections.length} sections.`,
        );
    }

    const stashValidationResult = validateAndResolveStashedItems(stash, initialItems);

    if (!isEmpty(stashValidationResult.missing)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to use non-existing stashes. Identifiers of missing stashes: ${stashValidationResult.missing.join(
                ", ",
            )}`,
        );
    }

    return stashValidationResult;
}

export function* addLayoutSectionHandler(
    ctx: DashboardContext,
    cmd: AddLayoutSection,
): SagaIterator<DashboardLayoutSectionAdded> {
    const commandCtx: AddLayoutSectionContext = {
        ctx,
        cmd,
        layout: yield select(selectLayout),
        stash: yield select(selectStash),
        availableInsights: yield select(selectInsightsMap),
    };

    const stashValidationResult = validateAndResolveItems(commandCtx);
    const {
        payload: { index, initialHeader, autoResolveDateFilterDataset },
    } = cmd;

    const normalizationResult: SagaReturnType<typeof validateAndNormalizeWidgetItems> = yield call(
        validateAndNormalizeWidgetItems,
        ctx,
        stashValidationResult,
        cmd,
    );

    const itemsToAdd: SagaReturnType<typeof validateAndResolveItemFilterSettings> = yield call(
        validateAndResolveItemFilterSettings,
        ctx,
        cmd,
        normalizationResult,
        autoResolveDateFilterDataset,
    );

    const section: ExtendedDashboardLayoutSection = {
        type: "IDashboardLayoutSection",
        header: initialHeader,
        items: itemsToAdd,
    };

    yield put(
        batchActions([
            insightsActions.addInsights(normalizationResult.resolvedInsights.loaded),
            layoutActions.addSection({
                section,
                usedStashes: stashValidationResult.existing,
                index,
                undo: {
                    cmd,
                },
            }),
        ]),
    );

    return layoutSectionAdded(
        ctx,
        section,
        resolveIndexOfNewItem(commandCtx.layout.sections, index),
        cmd.correlationId,
    );
}
