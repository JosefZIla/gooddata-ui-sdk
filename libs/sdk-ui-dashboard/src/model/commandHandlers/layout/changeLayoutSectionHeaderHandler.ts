// (C) 2021 GoodData Corporation
import { DashboardContext } from "../../types/commonTypes";
import { ChangeLayoutSectionHeader } from "../../commands";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import { commandRejected } from "../../events/general";

export function* changeLayoutSectionHeaderHandler(ctx: DashboardContext, cmd: ChangeLayoutSectionHeader) {
    yield dispatchDashboardEvent(commandRejected(ctx, cmd.correlationId));
}
