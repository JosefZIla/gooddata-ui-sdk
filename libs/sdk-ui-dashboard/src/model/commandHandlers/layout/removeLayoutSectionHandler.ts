// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { RemoveLayoutSection } from "../../commands";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import { commandRejected } from "../../events/general";

export function* removeLayoutSectionHandler(
    ctx: DashboardContext,
    cmd: RemoveLayoutSection,
): SagaIterator<void> {
    yield dispatchDashboardEvent(commandRejected(ctx, cmd.correlationId));
}
