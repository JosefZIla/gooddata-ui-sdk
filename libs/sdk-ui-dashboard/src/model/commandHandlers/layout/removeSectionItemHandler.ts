// (C) 2021 GoodData Corporation
import { DashboardContext } from "../../types/commonTypes";
import { RemoveSectionItem } from "../../commands";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import { commandRejected } from "../../events/general";

export function* removeSectionItemHandler(ctx: DashboardContext, cmd: RemoveSectionItem) {
    yield dispatchDashboardEvent(commandRejected(ctx, cmd.correlationId));
}
