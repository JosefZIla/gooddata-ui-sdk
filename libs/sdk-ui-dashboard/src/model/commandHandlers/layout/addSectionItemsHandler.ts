// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { AddSectionItems } from "../../commands";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import { invalidArgumentsProvided } from "../../events/general";
import { selectLayout, selectStash } from "../../state/layout/layoutSelectors";
import { put, select } from "redux-saga/effects";
import { validateItemPlacement, validateSectionExists } from "./validation/layoutValidation";
import { ExtendedDashboardLayoutSection } from "../../types/layoutTypes";
import { validateAndResolveStashedItems } from "./validation/stashValidation";
import isEmpty from "lodash/isEmpty";
import { layoutActions } from "../../state/layout";
import { layoutSectionItemsAdded } from "../../events/layout";

// TODO: this needs to handle calculation of the date dataset to use for the items
export function* addSectionItemsHandler(ctx: DashboardContext, cmd: AddSectionItems): SagaIterator<void> {
    const layout: ReturnType<typeof selectLayout> = yield select(selectLayout);
    const stash: ReturnType<typeof selectStash> = yield select(selectStash);

    const { items, itemIndex, sectionIndex } = cmd.payload;

    if (!validateSectionExists(layout, sectionIndex)) {
        return yield dispatchDashboardEvent(
            invalidArgumentsProvided(
                ctx,
                `Attempting to add items to non-existing layout section at index ${sectionIndex}.`,
                cmd.correlationId,
            ),
        );
    }

    const section: ExtendedDashboardLayoutSection = layout.sections[sectionIndex];

    if (!validateItemPlacement(section, itemIndex)) {
        return yield dispatchDashboardEvent(
            invalidArgumentsProvided(
                ctx,
                `Attempting to insert new item at wrong index ${itemIndex}. There are currently ${section.items.length} items.`,
                cmd.correlationId,
            ),
        );
    }

    const stashValidationResult = validateAndResolveStashedItems(stash, items);

    if (!isEmpty(stashValidationResult.missing)) {
        return yield dispatchDashboardEvent(
            invalidArgumentsProvided(
                ctx,
                `Attempting to use non-existing stashes. Identifiers of missing stashes: ${stashValidationResult.missing.join(
                    ", ",
                )}`,
                cmd.correlationId,
            ),
        );
    }

    yield put(
        layoutActions.addSectionItems({
            sectionIndex,
            itemIndex,
            items: stashValidationResult.resolved,
            usedStashes: stashValidationResult.existing,
            undo: {
                cmd,
            },
        }),
    );

    yield dispatchDashboardEvent(
        layoutSectionItemsAdded(
            ctx,
            sectionIndex,
            itemIndex < 0 ? section.items.length : itemIndex,
            stashValidationResult.resolved,
            stashValidationResult.existing,
            cmd.correlationId,
        ),
    );
}
