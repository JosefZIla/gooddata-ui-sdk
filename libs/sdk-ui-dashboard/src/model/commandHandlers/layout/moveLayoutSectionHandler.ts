// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { MoveLayoutSection } from "../../commands";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import { invalidArgumentsProvided } from "../../events/general";
import { selectLayout } from "../../state/layout/layoutSelectors";
import { put, select } from "redux-saga/effects";
import { validateSectionExists, validateSectionPlacement } from "./validation/layoutValidation";
import { layoutActions } from "../../state/layout";
import { layoutSectionMoved } from "../../events/layout";
import { resolveRelativeIndex } from "../../utils/arrayOps";

export function* moveLayoutSectionHandler(ctx: DashboardContext, cmd: MoveLayoutSection): SagaIterator<void> {
    const layout: ReturnType<typeof selectLayout> = yield select(selectLayout);
    const { sectionIndex, toIndex } = cmd.payload;

    if (!validateSectionExists(layout, sectionIndex)) {
        return yield dispatchDashboardEvent(
            invalidArgumentsProvided(
                ctx,
                `Attempting to move non-existent section from index ${sectionIndex}. There are only ${layout.sections.length} sections.`,
                cmd.correlationId,
            ),
        );
    }

    if (!validateSectionPlacement(layout, toIndex)) {
        return yield dispatchDashboardEvent(
            invalidArgumentsProvided(
                ctx,
                `Attempting to move section to a wrong index ${toIndex}. There are currently ${layout.sections.length} sections.`,
                cmd.correlationId,
            ),
        );
    }

    yield put(
        layoutActions.moveSection({
            sectionIndex,
            toIndex,
            undo: {
                cmd,
            },
        }),
    );

    const section = layout.sections[sectionIndex];
    const absoluteIndex = resolveRelativeIndex(layout.sections, toIndex);

    yield dispatchDashboardEvent(
        layoutSectionMoved(ctx, section, sectionIndex, absoluteIndex, cmd.correlationId),
    );
}
