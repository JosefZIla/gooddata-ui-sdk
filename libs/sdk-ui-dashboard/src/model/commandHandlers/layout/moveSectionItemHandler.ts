// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes";
import { MoveSectionItem } from "../../commands";
import { invalidArgumentsProvided } from "../../events/general";
import { selectLayout } from "../../state/layout/layoutSelectors";
import { put, select } from "redux-saga/effects";
import {
    validateItemExists,
    validateItemPlacement,
    validateSectionExists,
    validateSectionPlacement,
} from "./validation/layoutValidation";
import { layoutActions } from "../../state/layout";
import { DashboardLayoutSectionItemMoved, layoutSectionItemMoved } from "../../events/layout";
import { resolveIndexOfNewItem, resolveRelativeIndex } from "../../utils/arrayOps";

type MoveSectionItemContext = {
    readonly ctx: DashboardContext;
    readonly cmd: MoveSectionItem;
    readonly layout: ReturnType<typeof selectLayout>;
};

function validateAndResolve(commandCtx: MoveSectionItemContext) {
    const {
        ctx,
        layout,
        cmd: {
            payload: { sectionIndex, toSectionIndex, itemIndex, toItemIndex },
        },
    } = commandCtx;

    if (!validateSectionExists(layout, sectionIndex)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to move item from non-existent section at ${sectionIndex}. There are only ${layout.sections.length} sections.`,
        );
    }

    const fromSection = layout.sections[sectionIndex];

    if (!validateItemExists(fromSection, itemIndex)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to move non-existent item from index ${itemIndex}. There are only ${fromSection.items.length} items.`,
        );
    }

    const itemToMove = fromSection.items[itemIndex];

    if (!validateSectionPlacement(layout, toSectionIndex)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to move item to a wrong section at index ${toSectionIndex}. There are currently ${layout.sections.length} sections.`,
        );
    }

    const targetSectionIndex = resolveRelativeIndex(layout.sections, toSectionIndex);
    const targetSection = layout.sections[targetSectionIndex];

    if (!validateItemPlacement(targetSection, toItemIndex)) {
        throw invalidArgumentsProvided(
            ctx,
            commandCtx.cmd,
            `Attempting to move item to a wrong location at index ${toItemIndex}. Target section has ${targetSection.items.length} items.`,
        );
    }

    let targetItemIndex = 0;

    if (sectionIndex === targetSectionIndex) {
        targetItemIndex = resolveRelativeIndex(targetSection.items, toItemIndex);

        if (itemIndex === targetItemIndex) {
            throw invalidArgumentsProvided(
                ctx,
                commandCtx.cmd,
                `Attempting to move item to a same place where it already resides ${toItemIndex}.`,
            );
        }
    } else {
        targetItemIndex = resolveIndexOfNewItem(targetSection.items, toItemIndex);
    }

    return {
        targetSectionIndex,
        targetItemIndex,
        itemToMove,
    };
}

export function* moveSectionItemHandler(
    ctx: DashboardContext,
    cmd: MoveSectionItem,
): SagaIterator<DashboardLayoutSectionItemMoved> {
    const commandCtx: MoveSectionItemContext = {
        ctx,
        cmd,
        layout: yield select(selectLayout),
    };

    const { targetSectionIndex, targetItemIndex, itemToMove } = validateAndResolve(commandCtx);
    const { itemIndex, sectionIndex } = cmd.payload;

    yield put(
        layoutActions.moveSectionItem({
            sectionIndex,
            itemIndex,
            toSectionIndex: targetSectionIndex,
            toItemIndex: targetItemIndex,
            undo: {
                cmd,
            },
        }),
    );

    return layoutSectionItemMoved(
        ctx,
        itemToMove,
        sectionIndex,
        targetSectionIndex,
        itemIndex,
        targetItemIndex,
        cmd.correlationId,
    );
}
