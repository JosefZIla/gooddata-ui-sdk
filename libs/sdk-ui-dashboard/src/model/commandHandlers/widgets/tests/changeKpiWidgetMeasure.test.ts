// (C) 2021 GoodData Corporation
import {
    ComplexDashboardIdentifier,
    ComplexDashboardWidgets,
    MockAvailabilityWithDifferentRelevance,
    TestCorrelation,
    TimelineDateDatasetRef,
} from "../../../tests/Dashboard.fixtures";
import { DashboardTester, preloadedTesterFactory } from "../../../tests/DashboardTester";
import { changeKpiWidgetMeasure, loadDashboard } from "../../../commands";
import { ReferenceLdm } from "@gooddata/reference-workspace";
import { measureItem, uriRef } from "@gooddata/sdk-model";
import { DashboardCommandFailed, DashboardKpiWidgetMeasureChanged } from "../../../events";
import { selectWidgetByRef } from "../../../state/layout/layoutSelectors";

describe("change KPI widget measure handler", () => {
    describe("for any KPI", () => {
        const WidgetWithNoFilters = ComplexDashboardWidgets.FirstSection.ThirdKpi;

        let Tester: DashboardTester;
        beforeEach(preloadedTesterFactory((tester) => (Tester = tester), ComplexDashboardIdentifier));

        it("should replace measure and keep existing header", async () => {
            const event: DashboardKpiWidgetMeasureChanged = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(WidgetWithNoFilters.ref, measureItem(ReferenceLdm.Won)),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(event.payload.measure.title).toEqual("Won");
            expect(event.payload.kpiWidget.title).toEqual("Probability All Time");
            expect(event.payload.header).toBeUndefined();
            expect(event.payload.kpiWidget.kpi.metric).toEqual(event.payload.measure.ref);
        });

        it("should replace measure and switch widget title to measure title", async () => {
            const event: DashboardKpiWidgetMeasureChanged = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(WidgetWithNoFilters.ref, measureItem(ReferenceLdm.Won), "from-title"),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(event.payload.measure.title).toEqual("Won");
            expect(event.payload.kpiWidget.title).toEqual("Won");
            expect(event.payload.header).toEqual({ title: "Won" });
            expect(event.payload.kpiWidget.kpi.metric).toEqual(event.payload.measure.ref);
        });

        it("should replace measure and switch widget title to custom value", async () => {
            const event: DashboardKpiWidgetMeasureChanged = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(WidgetWithNoFilters.ref, measureItem(ReferenceLdm.Won), {
                    title: "Custom Title",
                }),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(event.payload.measure.title).toEqual("Won");
            expect(event.payload.kpiWidget.title).toEqual("Custom Title");
            expect(event.payload.header).toEqual({ title: "Custom Title" });
            expect(event.payload.kpiWidget.kpi.metric).toEqual(event.payload.measure.ref);
        });

        it("should fail if widget does not exist", async () => {
            const event: DashboardCommandFailed = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(
                    uriRef("does-not-exist"),
                    measureItem(ReferenceLdm.Won),
                    { title: "Custom Title" },
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });

        it("should fail if measure does not exist", async () => {
            const event: DashboardCommandFailed = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(
                    WidgetWithNoFilters.ref,
                    uriRef("does-not-exist"),
                    { title: "Custom Title" },
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });

        it("should emit correct events", async () => {
            await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(
                    WidgetWithNoFilters.ref,
                    measureItem(ReferenceLdm.Won),
                    "from-title",
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });

    describe("for KPI with date dataset", () => {
        const WidgetWithDateDataset = ComplexDashboardWidgets.FirstSection.LastKpi;

        it("should keep existing date dataset if it is available for the new metric", async () => {
            const Tester = DashboardTester.forRecording(ComplexDashboardIdentifier);
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.resetMonitors();

            const originalWidget = selectWidgetByRef(WidgetWithDateDataset.ref)(Tester.state());
            const event: DashboardKpiWidgetMeasureChanged = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(WidgetWithDateDataset.ref, measureItem(ReferenceLdm.Won)),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(event.payload.kpiWidget.dateDataSet).toEqual(originalWidget!.dateDataSet);
        });

        it("should switch to most relevant date dataset if existing setting is not valid", async () => {
            const Tester = DashboardTester.forRecording(
                ComplexDashboardIdentifier,
                {},
                {
                    catalogAvailability: {
                        availableDateDatasets: MockAvailabilityWithDifferentRelevance,
                    },
                },
            );
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.resetMonitors();

            const originalWidget = selectWidgetByRef(WidgetWithDateDataset.ref)(Tester.state());
            const event: DashboardKpiWidgetMeasureChanged = await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(WidgetWithDateDataset.ref, measureItem(ReferenceLdm.Won)),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(event.payload.kpiWidget.dateDataSet).not.toEqual(originalWidget!.dateDataSet);
            // this is URI of the Timeline date dataset which the mocks used above make the most relevant
            // the Timeline date dataset is part of the reference workspace catalog
            expect(event.payload.kpiWidget.dateDataSet).toEqual(TimelineDateDatasetRef);
        });

        it("should emit correct events", async () => {
            const Tester = DashboardTester.forRecording(ComplexDashboardIdentifier);
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.resetMonitors();

            await Tester.dispatchAndWaitFor(
                changeKpiWidgetMeasure(
                    WidgetWithDateDataset.ref,
                    measureItem(ReferenceLdm.Won),
                    undefined,
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });
});
