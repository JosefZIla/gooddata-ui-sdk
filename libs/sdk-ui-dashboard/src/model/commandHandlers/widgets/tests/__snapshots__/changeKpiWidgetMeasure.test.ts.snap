// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change KPI widget measure handler for KPI with date dataset should emit correct events 1`] = `
Array [
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.QUERY.STARTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.QUERY.COMPLETED",
  },
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
  },
]
`;

exports[`change KPI widget measure handler for any KPI should emit correct events 1`] = `
Array [
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.KPI_WIDGET.MEASURE_CHANGED",
  },
]
`;
